File: backend/app/schemas/fitness.py
"""Fitness schemas"""

from typing import Optional, List
from uuid import UUID
from datetime import date, datetime
from pydantic import BaseModel, Field


class ExerciseBase(BaseModel):
    """Base exercise schema"""
    exercise_order: int = Field(..., ge=1)
    name: str
    description: Optional[str] = None
    sets: Optional[int] = Field(None, ge=1)
    reps: Optional[int] = Field(None, ge=1)
    duration_seconds: Optional[int] = Field(None, ge=0)
    rest_time_seconds: Optional[int] = Field(None, ge=0)
    muscle_groups: Optional[List[str]] = None
    equipment_required: Optional[List[str]] = None
    instructions: Optional[str] = None
    video_url: Optional[str] = None
    form_cues: Optional[List[str]] = None
    difficulty_level: Optional[int] = Field(None, ge=1, le=10)


class ExerciseResponse(ExerciseBase):
    """Exercise response schema"""
    id: UUID
    workout_id: UUID
    created_at: datetime
    
    model_config = {"from_attributes": True}


class WorkoutBase(BaseModel):
    """Base workout schema"""
    day_of_week: int = Field(..., ge=0, le=6)
    workout_type: str
    name: str
    description: Optional[str] = None
    duration_minutes: int = Field(..., gt=0)
    estimated_calories: Optional[int] = Field(None, ge=0)
    intensity_level: Optional[int] = Field(None, ge=1, le=10)


class WorkoutResponse(WorkoutBase):
    """Workout response schema"""
    id: UUID
    plan_id: UUID
    created_at: datetime
    exercises: List[ExerciseResponse] = []
    
    model_config = {"from_attributes": True}


class FitnessPlanBase(BaseModel):
    """Base fitness plan schema"""
    week_start: date
    week_end: date
    difficulty_level: Optional[int] = Field(None, ge=1, le=10)
    focus_areas: Optional[List[str]] = None
    estimated_calories_burn: Optional[int] = Field(None, ge=0)


class FitnessPlanCreate(FitnessPlanBase):
    """Fitness plan creation schema"""
    workouts: Optional[List[WorkoutBase]] = None


class FitnessPlanResponse(FitnessPlanBase):
    """Fitness plan response schema"""
    id: UUID
    user_id: UUID
    created_by_ai: bool
    adaptation_reason: Optional[str] = None
    is_active: bool
    version: int
    created_at: datetime
    workouts: List[WorkoutResponse] = []
    
    model_config = {"from_attributes": True}


class FitnessPlanListResponse(BaseModel):
    """List of fitness plans"""
    plans: List[FitnessPlanResponse]
    total: int


class WorkoutLogBase(BaseModel):
    """Base workout log schema"""
    workout_date: date
    workout_name: Optional[str] = None
    duration_minutes: int = Field(..., gt=0)
    calories_burned: Optional[int] = Field(None, ge=0)
    perceived_exertion: Optional[int] = Field(None, ge=1, le=10)
    mood_after: Optional[int] = Field(None, ge=1, le=10)
    completed: bool = True
    completion_percentage: Optional[int] = Field(100, ge=0, le=100)
    notes: Optional[str] = None


class WorkoutLogCreate(WorkoutLogBase):
    """Workout log creation schema"""
    workout_id: Optional[UUID] = None


class WorkoutLogResponse(WorkoutLogBase):
    """Workout log response schema"""
    id: UUID
    user_id: UUID
    workout_id: Optional[UUID] = None
    logged_at: datetime
    
    model_config = {"from_attributes": True}


# Additional schemas for API routes
class WorkoutPlanResponse(BaseModel):
    """Simplified workout plan response for API"""
    id: str
    user_id: str
    week_start: str
    workouts: List[dict]
    difficulty_level: int
    created_by_ai: bool
    adaptation_reason: Optional[str] = None

File: flutter_app/lib/models/fitness.dart
import 'package:json_annotation/json_annotation.dart';

part 'fitness.g.dart';

@JsonSerializable()
class WorkoutPlan {
  final String id;
  @JsonKey(name: 'user_id')
  final String userId;
  @JsonKey(name: 'week_start')
  final String weekStart;
  @JsonKey(name: 'difficulty_level')
  final int? difficultyLevel;
  @JsonKey(name: 'focus_areas')
  final List<String>? focusAreas;
  @JsonKey(name: 'estimated_calories_burn')
  final int? estimatedCaloriesBurn;
  final List<DailyWorkout> workouts;
  @JsonKey(name: 'created_by_ai')
  final bool createdByAi;
  @JsonKey(name: 'adaptation_reason')
  final String? adaptationReason;

  WorkoutPlan({
    required this.id,
    required this.userId,
    required this.weekStart,
    this.difficultyLevel,
    this.focusAreas,
    this.estimatedCaloriesBurn,
    required this.workouts,
    required this.createdByAi,
    this.adaptationReason,
  });

  factory WorkoutPlan.fromJson(Map<String, dynamic> json) => _$WorkoutPlanFromJson(json);
  Map<String, dynamic> toJson() => _$WorkoutPlanToJson(this);
}

@JsonSerializable()
class DailyWorkout {
  final String day;
  final List<Workout> workouts;

  DailyWorkout({
    required this.day,
    required this.workouts,
  });

  factory DailyWorkout.fromJson(Map<String, dynamic> json) => _$DailyWorkoutFromJson(json);
  Map<String, dynamic> toJson() => _$DailyWorkoutToJson(this);
}

@JsonSerializable()
class Workout {
  final String id;
  final String name;
  final String? description;
  @JsonKey(name: 'duration_minutes')
  final int durationMinutes;
  @JsonKey(name: 'estimated_calories')
  final int? estimatedCalories;
  @JsonKey(name: 'intensity_level')
  final int? intensityLevel;
  final List<Exercise> exercises;

  Workout({
    required this.id,
    required this.name,
    this.description,
    required this.durationMinutes,
    this.estimatedCalories,
    this.intensityLevel,
    required this.exercises,
  });

  factory Workout.fromJson(Map<String, dynamic> json) => _$WorkoutFromJson(json);
  Map<String, dynamic> toJson() => _$WorkoutToJson(this);
}

@JsonSerializable()
class Exercise {
  final String id;
  final String name;
  final String? description;
  final int? sets;
  final int? reps;
  @JsonKey(name: 'duration_seconds')
  final int? durationSeconds;
  @JsonKey(name: 'rest_time_seconds')
  final int? restTimeSeconds;
  @JsonKey(name: 'muscle_groups')
  final List<String>? muscleGroups;
  @JsonKey(name: 'equipment_required')
  final List<String>? equipmentRequired;
  final String? instructions;
  @JsonKey(name: 'video_url')
  final String? videoUrl;
  @JsonKey(name: 'form_cues')
  final List<String>? formCues;

  Exercise({
    required this.id,
    required this.name,
    this.description,
    this.sets,
    this.reps,
    this.durationSeconds,
    this.restTimeSeconds,
    this.muscleGroups,
    this.equipmentRequired,
    this.instructions,
    this.videoUrl,
    this.formCues,
  });

  factory Exercise.fromJson(Map<String, dynamic> json) => _$ExerciseFromJson(json);
  Map<String, dynamic> toJson() => _$ExerciseToJson(this);
}